src/goto-instrument/contracts/dynamic-frames/dfcc_instrument.h

```diff
@@ -34,6 +41,48 @@ class cfg_infot;
 class dfcc_libraryt;
 class dfcc_utilst;
 
+/// A local pragma used to keep track instructions in loops.
+const char CONTRACT_PRAGMA_loop_id[] = "contracts:loop-id";
+const char CONTRACT_PRAGMA_loop_body[] = "contracts:loop-id-body";
+const char CONTRACT_PRAGMA_loop_head[] = "contracts:loop-id-head";
+const char CONTRACT_PRAGMA_loop_latch[] = "contracts:loop-id-latch";
+const char CONTRACT_PRAGMA_loop_exiting[] = "contracts:loop-id-exiting";
+
+const char empty_pragma[] = "";
+
+// A graph node type that stores information about a loop.
+// We create a DAG representing nesting of various loops in goto_function,
+// sort them topologically, and instrument them in the top-sorted order.
+//
+// The goal here is not avoid explicit "subset checks" on nested write sets.
+// When instrumenting in a top-sorted order,
+// the outer loop would no longer observe the inner loop's write set,
+// but only corresponding `havoc` statements,
+// which can be instrumented in the usual way to achieve a "subset check".
+struct loop_graph_nodet : public graph_nodet<empty_edget>
+{
+  natural_loops_mutablet::loopt &content;
+  goto_programt::targett head_target, latch_target;
+  exprt assigns_clause, invariant, decreases_clause;
+  symbolt write_set_symbol;
+
+  loop_graph_nodet(
+    natural_loops_mutablet::loopt &loop,
+    const goto_programt::targett head,
+    const goto_programt::targett latch,
+    const exprt &assigns,
```

Comment:
isn't this supposed to be a vector/list  of `conditional_target_group_exprt`

Reply:
Can we encapsulate the GOTO converter instance here ? Are these expressions supposed to be already GOTO converted or not ? When does goto_conversion happen for the loop contract expressions ?